### DISCUSSION TITLE
howto have rust-analyzer work with different Rust projects

### DISCUSSION THREAD

**Discussion by tshepang** [helix-editor/helix#5889] [Q&A] [answered]
I find that I have to `lsp-restart` when I switch to a different project. How do I avoid this? I assume that rust-analyzer can't do more than a single project in 1 process, so it would have to span multiple?

**Comment by pascalkuthe** [comment, marked as answer]
The reason this doesn't work is a bit nuanced. Rust analyzer does support multiple projects roots, however there are some problems with that:

* Helix automatically detects project roots that means it can not know all projects in a workspace ahead of time
* On master helix simply uses the root of the first file opened for a language (and root discovery is also limited). 
* #5748 improves on this by rediscovering the LSP root and notifying.the sever of the nee roots when necessary. For LSPs that support multi workspaces this is really seamless, for LSPs that don't, it automatically spawns multiple LS instances instead.
* In theory rust-analyzer does support multiple workspaces internally but that doesn't actually work because it doesn't implement all the relevant parts of the lsp spec I fixed that in rust-lang/rust-analyzer#14098 which was already merged. 

So TLDR #5748 fixes that problem, which you could testdrive. It doesn't work aswell as it could right now, because rust-analyzer was missing something.
For now you can easily build RA from source (you.need the nightly toolchain but then a simple `cargo install --path crates/rust-analyzer` will do) but the fix should be included in the release next Monday

**Comment by jyn514** [comment]
hmm, i am not sure what i am doing wrong, but this answer does not work for me. i have two cargo projects in adjacent directories (`~/src/helix` and `~/src/cargo`). running `cd cargo && hx src/bin/cargo/main.rs` and then `:open ~/src/helix/helix-term/src/application.rs` only has an LSP running for cargo; and vice versa if i open the helix file first.
